core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * When generating new DEKs for OutputFiles, this class will first attempt to have the KMS generate a new key. If the
+ * KMS does not support key generation a new key will be produced by pulling bytes from a {@link SecureRandom} on the
+ * JVM writing the file.
+ */
+public class EnvelopeEncryptionManager implements EncryptionManager {
+  private final EnvelopeConfiguration dataEncryptionConfig;
+  private final boolean nativeFormatEncryption;
+  private final KmsClient kmsClient;
+  private final int dataKeyLength;
+  private final boolean kmsGeneratedKeys;
+
+  private transient volatile SecureRandom workerRNG = null;
+
+  public static final String encryptionConfigMismatchMessagePrefix =
+      "Cannot create encryption manager - detected mismatch between client encryption properties and table " +
+          "encryption properties. Mismatch can be caused by changing encryption properties after table creation " +
+          "(currently not supported), or by malicious tampering with table properties kept in metadata.json in " +
+          "storage. Mismatch details: ";
+
+  /**
+   * @param nativeFormatEncryption true for native encryption inside file formats (Parquet or ORC), false for generic
+   *                               encryption of file streams
+   * @param dataEncryptionConfig configuration for encryption of data files. Future versions will add configuration for
+   *                             encryption of metadata files
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   * @param dataKeyLength If DEKs are not generated by KMS, they will be randomly generated by Iceberg using this length
+   */
+  public EnvelopeEncryptionManager(boolean nativeFormatEncryption,  EnvelopeConfiguration dataEncryptionConfig,
+      KmsClient kmsClient, int dataKeyLength) {
+    if (!nativeFormatEncryption) {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently only supports encryption " +
+          "provided by the underlying file format.");
+    }
+    this.nativeFormatEncryption = nativeFormatEncryption;
+    Preconditions.checkNotNull(dataEncryptionConfig,
+        "Cannot create EnvelopeEncryptionManager because data encryption config is not passed");
+    this.dataEncryptionConfig = dataEncryptionConfig;
+    Preconditions.checkNotNull(dataEncryptionConfig.kekId(),
+        "Cannot create EnvelopeEncryptionManager because table key encryption key ID is not specified");
+    Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.kmsClient = kmsClient;
+    this.dataKeyLength = dataKeyLength;
+    this.kmsGeneratedKeys = kmsClient.supportsKeyGeneration();
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
```

Comment:
Same here, should generate keys only when confirming that encryption is required.